File: oltrack-ts/src/normalize.ts
import type { Shipment, TrackingEvent, UlscCode } from "./types.js";

/**
 * Build a TrackingEvent dict from individual fields.
 *
 * The caller MUST supply oltsCode — unlike the Python implementation,
 * the TS SDK doesn't bundle the CSV mappings inline (they live in the
 * repo's mappings/*.csv and would be a build-time dependency).
 *
 * Adapters in `adapters/*` know their carrier's raw_code → olts_code
 * mapping at compile time.
 */
export function event(input: {
  eventTime: string;
  oltsCode: UlscCode;
  carrierCode?: string;
  carrierEventCode?: string;
  description?: string | Record<string, string>;
  location?: TrackingEvent["location"];
  operator?: TrackingEvent["operator"];
  pieceId?: string;
  transport?: TrackingEvent["transport"];
  isLogicalEvent?: boolean;
  notes?: string;
  metadata?: Record<string, unknown>;
}): TrackingEvent {
  const ev: TrackingEvent = {
    eventTime: input.eventTime,
    oltsCode: input.oltsCode,
  };
  if (input.carrierCode) ev.carrierCode = input.carrierCode;
  if (input.carrierEventCode) ev.carrierEventCode = input.carrierEventCode;
  if (input.description) ev.description = input.description;
  if (input.location && Object.keys(input.location).length > 0) ev.location = input.location;
  if (input.operator && Object.keys(input.operator).length > 0) ev.operator = input.operator;
  if (input.pieceId) ev.pieceId = input.pieceId;
  if (input.transport) ev.transport = input.transport;
  if (input.isLogicalEvent) ev.isLogicalEvent = true;
  if (input.notes) ev.notes = input.notes;
  if (input.metadata && Object.keys(input.metadata).length > 0) ev.metadata = input.metadata;
  return ev;
}

/**
 * Build a Shipment dict wrapping a list of events.
 * `currentStatus` auto-derives from last event's oltsCode if not explicitly set.
 */
export function shipment(input: {
  trackingNumber: string;
  events: TrackingEvent[];
  carrierCode?: string;
  service?: string;
  currentStatus?: UlscCode;
  origin?: Shipment["origin"];
  destination?: Shipment["destination"];
  estimatedDelivery?: Shipment["estimatedDelivery"];
  actualDelivery?: string;
  pieces?: Shipment["pieces"];
  weight?: Shipment["weight"];
  dimensions?: Shipment["dimensions"];
  declaredValue?: Shipment["declaredValue"];
  isReturn?: boolean;
  metadata?: Record<string, unknown>;
}): Shipment {
  const ship: Shipment = {
    trackingNumber: input.trackingNumber,
    events: input.events,
  };
  if (input.carrierCode) ship.carrierCode = input.carrierCode;
  if (input.service) ship.service = input.service;
  // Auto-derive currentStatus from last event if not explicitly set
  if (input.currentStatus) {
    ship.currentStatus = input.currentStatus;
  } else if (input.events.length > 0) {
    ship.currentStatus = input.events[input.events.length - 1]!.oltsCode;
  }
  if (input.origin) ship.origin = input.origin;
  if (input.destination) ship.destination = input.destination;
  if (input.estimatedDelivery) ship.estimatedDelivery = input.estimatedDelivery;
  if (input.actualDelivery) ship.actualDelivery = input.actualDelivery;
  if (input.pieces && input.pieces.length > 0) ship.pieces = input.pieces;
  if (input.weight) ship.weight = input.weight;
  if (input.dimensions) ship.dimensions = input.dimensions;
  if (input.declaredValue) ship.declaredValue = input.declaredValue;
  if (input.isReturn) ship.isReturn = true;
  if (input.metadata && Object.keys(input.metadata).length > 0) ship.metadata = input.metadata;
  return ship;
}

File: oltrack-py/src/oltrack/normalize.py
"""Top-level normalization entry points.

Adapters in :mod:`oltrack.adapters` provide carrier-specific parsing of raw
API responses into TrackingEvent dicts. This module provides a single-event
helper for callers who have already done the parsing.
"""

from __future__ import annotations

from typing import Any, Optional

from oltrack.mappings import lookup


def event(
    *,
    carrier: str,
    carrier_event_code: str,
    event_time: str,
    description: Optional[Any] = None,
    location: Optional[dict[str, Any]] = None,
    operator: Optional[dict[str, Any]] = None,
    piece_id: Optional[str] = None,
    transport: Optional[dict[str, Any]] = None,
    is_logical_event: bool = False,
    notes: Optional[str] = None,
    metadata: Optional[dict[str, Any]] = None,
    olts_code: Optional[str] = None,
) -> dict[str, Any]:
    """Build a single OLTS TrackingEvent dict.

    If ``olts_code`` is not supplied, looks it up from the carrier mapping.

    Args:
        carrier: Carrier slug (e.g. 'sf', 'dhl').
        carrier_event_code: Raw event code as returned by the carrier API.
        event_time: ISO 8601 timestamp with timezone offset.
        olts_code: Optional explicit ULSC code; if omitted, derived from
            carrier+carrier_event_code via mappings/.

    Returns:
        A dict shaped to the OLTS v0.2 TrackingEvent JSON Schema. Omits any
        optional fields the caller didn't provide.
    """
    if olts_code is None:
        olts_code = lookup(carrier, carrier_event_code)

    out: dict[str, Any] = {
        "eventTime": event_time,
        "carrierCode": carrier,
        "carrierEventCode": carrier_event_code,
        "oltsCode": olts_code,
    }
    if description is not None:
        out["description"] = description
    if location:
        out["location"] = location
    if operator:
        out["operator"] = operator
    if piece_id:
        out["pieceId"] = piece_id
    if transport:
        out["transport"] = transport
    if is_logical_event:
        out["isLogicalEvent"] = True
    if notes:
        out["notes"] = notes
    if metadata:
        out["metadata"] = metadata
    return out


def shipment(
    *,
    tracking_number: str,
    events: list[dict[str, Any]],
    carrier: Optional[str] = None,
    **extras: Any,
) -> dict[str, Any]:
    """Build a Shipment dict wrapping a list of normalized events.

    ``extras`` accepts any of the optional Shipment fields (service,
    currentStatus, origin, destination, pieces, weight, dimensions,
    declaredValue, isReturn, metadata).
    """
    out: dict[str, Any] = {
        "trackingNumber": tracking_number,
        "events": events,
    }
    if carrier:
        out["carrierCode"] = carrier
    # currentStatus auto-derive: take the last event's oltsCode
    if "currentStatus" not in extras and events:
        out["currentStatus"] = events[-1]["oltsCode"]
    for k in ("service", "currentStatus", "origin", "destination",
              "estimatedDelivery", "actualDelivery", "pieces",
              "weight", "dimensions", "declaredValue", "isReturn", "metadata"):
        if k in extras and extras[k] is not None:
            out[k] = extras[k]
    return out
